### DISCUSSION TITLE
Memory consumption with the merge function

### DISCUSSION THREAD

**Discussion by gil-c** [libvips/libvips#2284] [Q&A]
Hi,

Given the pipeline architecture of LibVips, I would have hoped that compositing a large image from small tiles (mosaicing) could have been achieved with a small memory footprint. Unfortunately, in practice the size taken in memory is about the total size of the tiles.

For now, the way I've found to limit the memory usage is by saving some medium size, intermediate images made with the merge function and then write the finale big image from there.

Is there a better solution?

Here is a script to demonstrate the memory consumption issue with merge:

```
import pyvips as vips
from pathlib import Path
import os


# tile size
width = 2000
height = 1500

# overlap joins by this many pixels
h_overlap = 100
v_overlap = 100

# number of tiles in mosaic
nb_cols = 40
nb_rows = 40


def join_left_right(tiles_row):
    row = tiles_row[0]
    for image in tiles_row[1:]:
        row = row.merge(image, 'horizontal', h_overlap - row.width, 0)
    return row


def join_top_bottom(rows):
    mos = rows[0]
    for row in rows[1:]:
        mos = mos.merge(row, 'vertical', 0, v_overlap - mos.height)
    return mos


# generate some tiles
tiles = [[vips.Image.black(width, height) for j in range(nb_cols)] for i in range(nb_rows)]

# write them on disk
tile_dir = Path(os.getcwd()).joinpath("tiles")
tile_dir.mkdir(exist_ok=True)

for i in range(len(tiles)):
    for j in range(len(tiles[i])):
        file_path = tile_dir.joinpath(f"{i:03}_{j:03}.tiff")
        if not file_path.exists():
            tiles[i][j].tiffsave(str(file_path))
        tiles[i][j] = vips.Image.tiffload(str(file_path))

# build the mosaic
rows = []
for tiles_row in tiles:
    rows.append(join_left_right(tiles_row))
mosaic = join_top_bottom(rows)

# save it
mosaic.tiffsave("output.tiff", tile=True, bigtiff=True)
```

Thanks!

**Comment by jcupitt** [comment]
Hi @gil-c,

The memory is mostly being used by the TIFF loader. When libvips opens a file, it has to convert it to a random-access format. For small files like these (3MB each), it simply loads them into memory. You have 1600 tiles, so that's 5GB used immediately.

You have some things you can do to address this. One improvement is to enable sequential mode -- this will hint to libvips that in this case it doesn't need to convert to a random-access format and it can just stream pixels off disc. 

I changed your code slightly to be:

```python
for i in range(len(tiles)):
    for j in range(len(tiles[i])):
        file_path = tile_dir.joinpath(f"{i:03}_{j:03}.tiff")
        if not file_path.exists():
            tiles[i][j].write_to_file(str(file_path))
        tiles[i][j] = vips.Image.new_from_file(str(file_path))
```

If I run it, I see:

```
$ /usr/bin/time -f %M:%s ./mosaic.py 
11412224:38
```

ie. 11gb of memory and 38s of elapsed time.

If I enable sequential mode:

```python
        tiles[i][j] = vips.Image.new_from_file(str(file_path), access="sequential")
```

I see:

```
$ /usr/bin/time -f %M:%e ./mosaic.py 
7765592:29.20
```

"only" 8GB and 29s.

Although it's now streaming pixels from the TIFF loader, the TIFF loader itself is quite heavy -- it includes a few caches. If you swap for VIPS format, memuse comes down more:

```python
        file_path = tile_dir.joinpath(f"{i:03}_{j:03}.v")
```

```
$ /usr/bin/time -f %M:%e ./mosaic.py 
1763428:17.22
```

1.7GB. 

Finally, that's running with a full threadpool, and libvips will be managing (for this PC) 32 copies of the pipeline. There's not actually that much parallelism here, so you can drop the thread count without hurting speed. For example:

```
$ VIPS_CONCURRENCY=1 /usr/bin/time -f %M:%e ./mosaic.py 
783516:21.22
```

700MB and 21s.

**Reply by jcupitt** [reply]
There's been some talk of changing the way loaders work so they can free memory more aggressively in cases like this, but it's not happened yet.

You may need to adjust your system to allow more files open at once. Linux systems usually default to 1024 files -- I have mine set to 65536. Windows has a hard limit of 2048.

**Reply by jcupitt** [reply]
I meant to add, there's a chapter in the manual explaining how libvips opens files, it might be helpful:

https://libvips.github.io/libvips/API/current/How-it-opens-files.md.html

**Reply by gil-c** [reply]
Thanks a lot John, as usual very rapid, concise and useful answers!

I was not aware of this 2048 max file limit on Windows. That will force me to keep the intermediate images mechanism then, since we are on Windows and we will have up to 100'000 fields of view to assemble in the future.

I will follow the other applicable advice.

Kind regards

Gilles

**Reply by jcupitt** [reply]
It's not really a windows limit (I think you can now have almost any number of HANDLEs open), it's a limit in implementation of open() in the standard windows C runtime.
